scripts/firedrake-install

```diff
@@ -0,0 +1,395 @@
+   installed to the specified location.
+
+The install location may be specified using the --global, --prefix, or --user
+options. If none of these options are provided, a local installation into
+./firedrake will be performed.""",
+                            formatter_class=RawDescriptionHelpFormatter)
+
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument("--global", action='store_true', dest="system",
+                       help="Install to the default global location on the current platform.")
+    # group.add_argument("--prefix", type=str, nargs=1,
+    #                   help="Install to subdirectories of the specified directory.")
+    group.add_argument("--user", action='store_true',
+                       help="Install using the current user's home directory as the prefix.")
+    parser.add_argument("--developer", action='store_true',
+                        help="Install in developer mode where the core packages are run from their source directories. Due to an upstream bug, petsc4py will not be installed in developer mode.")
+    parser.add_argument("--sudo", action='store_true',
+                        help="Use sudo when installing system dependencies and python packages. Regardless of the presence of this flag, sudo will never be used to install packages into $HOME or ./firedrake but sudo will always be used when installing system dependencies using apt.")
+
+    parser.add_argument("--log", action='store_true',
+                        help="Produce a verbose log of the installation process in firedrake-install.log. If you have problem running this script, please include this log in any bug report you file.")
+    parser.add_argument("--disable_ssh", action="store_true",
+                        help="Do not attempt to use ssh to clone git repositories: fall immediately back to https.")
+
+    args = parser.parse_args()
+    args.prefix = False  # Disabled as untested
+else:
+    parser = ArgumentParser(description="""Update this firedrake install to the latest versions of all packages.""",
+                            formatter_class=RawDescriptionHelpFormatter)
+    parser.add_argument("--rebuild", action="store_true",
+                        help="Rebuild all packages even if no new version is available. Usually petsc and petsc4py are only rebuilt if they change. All other packages are always rebuilt.")
+    parser.add_argument("--log", action='store_true',
+                        help="Produce a verbose log of the update process in firedrake-update.log. If you have problem running this script, please include this log in any bug report you file.")
+
+    args = parser.parse_args()
+    args.system = False
+    args.prefix = False
+    args.user = False
+    args.developer = False
+    args.sudo = False
+
+
+# Where are packages installed relative to --root?
+# This is a bit of a hack.
+v = sys.version_info
+sitepackages = "/usr/local/lib/python%d.%d/site-packages/" % (v.major, v.minor)
```

Comment:
It's a total mess. You'll also notice that --prefix installs are currently disabled because I haven't managed to make them work reliably.